File: greedy/max_sum_b_gt_a.cpp
/*
 * Problem: Max Sum of B where B[i] > A[i]
 * Difficulty: Medium
 * Pattern: Greedy / Multiset
 *
 * Given two integer arrays a and b of size n:
 * - You can sort b (but not a)
 * - Maximize the sum of all b[i] where b[i] > a[i]
 *
 * Approach:
 *   Sort b descending. For each b value, greedily pair it with the
 *   LARGEST a[i] it can strictly beat. This preserves smaller a values
 *   for future (smaller) b values to also win.
 *
 *   Use a multiset for O(log n) lookup of largest a < b_val.
 *
 * Time: O(n log n), Space: O(n)
 */

#include <bits/stdc++.h>
using namespace std;

long long maxSum(vector<int>& a, vector<int>& b) {
    sort(b.rbegin(), b.rend());
    multiset<int> ms(a.begin(), a.end());

    long long sum = 0;
    for (int bval : b) {
        auto it = ms.lower_bound(bval); // first a >= bval
        if (it != ms.begin()) {
            --it;                        // largest a < bval
            sum += bval;
            ms.erase(it);
        } else {
            //ms.erase(ms.begin());        // can't win, discard smallest
        }
    }
    return sum;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    cin >> n;
    vector<int> a(n), b(n);
    for (int& x : a) cin >> x;
    for (int& x : b) cin >> x;

    cout << maxSum(a, b) << "\n";
    return 0;
}

File: greedy/MaxSumBGtA.java
/*
 * Problem: Max Sum of B where B[i] > A[i]
 * Pattern: Greedy / TreeMap
 *
 * Sort b descending. For each b value, pair with largest a it can strictly beat.
 * Time: O(n log n), Space: O(n)
 */

import java.util.*;

public class MaxSumBGtA {
    public static long maxSum(List<Integer> a, List<Integer> b) {
        List<Integer> sortedB = new ArrayList<>(b);
        sortedB.sort(Collections.reverseOrder());

        TreeMap<Integer, Integer> map = new TreeMap<>();
        for (int x : a) map.merge(x, 1, Integer::sum);

        long sum = 0;
        for (int bval : sortedB) {
            Integer key = map.lowerKey(bval); // largest a strictly < bval
            if (key != null) {
                sum += bval;
                if (map.get(key) == 1) map.remove(key);
                else map.merge(key, -1, Integer::sum);
            } else {
                int first = map.firstKey();
                if (map.get(first) == 1) map.remove(first);
                else map.merge(first, -1, Integer::sum);
            }
        }
        return sum;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        List<Integer> a = new ArrayList<>(), b = new ArrayList<>();
        for (int i = 0; i < n; i++) a.add(sc.nextInt());
        for (int i = 0; i < n; i++) b.add(sc.nextInt());
        System.out.println(maxSum(a, b));
    }
}
